### PR TITLE
Add harfbuzz library

### PR DESCRIPTION
**Pull request by tpimh**
Again, I haven't yet tested if it works, so I am creating this PR as a draft. Just want to know if anything should be changed.

### CODE REVIEW

**Comment by tpimh** [comment]
Testing it now. It seems that `harfbuzzConfig.cmake` references `Threads::Threads`, and pspsdk doesn't provide `FindThreads.cmake`. There probably should be a workaround for it.

Anyway, my test code is [here](https://github.com/tpimh/jewels-SDL2-PSP-test/tree/freetype-harfbuzz) (based on a project by @sharkwouter), so if anyone wants to try and test this library, it should be a good start.

**Comment by sharkwouter** [comment]
So for threads we have pthreads, which is also used in C++ by threads. You can find the source here: https://github.com/pspdev/pthread-embedded/tree/platform_agnostic

**Comment by sharkwouter** [comment]
So, I installed both sdl2-ttf and harfbuzz from the automated build and this is the result I got:
```
wouter@wouter-gridscale:~/Personal/jewels-SDL2-PSP-test/psp$ psp-cmake ..
-- The C compiler identification is GNU 13.2.0
-- The CXX compiler identification is GNU 13.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /home/wouter/Personal/pspdev/bin/psp-gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /home/wouter/Personal/pspdev/bin/psp-g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PkgConfig: /home/wouter/Personal/pspdev/bin/psp-pkg-config (found version "0.29.2") 
-- Checking for one of the modules 'sdl2'
-- Found Freetype: /home/wouter/Personal/pspdev/psp/lib/libfreetype.a (found version "2.11.0") 
CMake Error at CMakeLists.txt:11 (find_package):
  By not providing "FindHarfbuzz.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Harfbuzz",
  but CMake did not find one.

  Could not find a package configuration file provided by "Harfbuzz" with any
  of the following names:

    HarfbuzzConfig.cmake
    harfbuzz-config.cmake

  Add the installation prefix of "Harfbuzz" to CMAKE_PREFIX_PATH or set
  "Harfbuzz_DIR" to a directory containing one of the above files.  If
  "Harfbuzz" provides a separate development package or SDK, be sure it has
  been installed.


-- Configuring incomplete, errors occurred!
See also "/home/wouter/Personal/jewels-SDL2-PSP-test/psp/CMakeFiles/CMakeOutput.log".

```

**Comment by tpimh** [comment]
My bad, for some reason on Ubuntu CMake is okay with "Harfbuzz" package, but for PSP it requires lowercase "harfbuzz". Not sure why CMake behaves differently, but I have updated CMakeLists.txt now.

**Comment by tpimh** [comment]
Two more things that are unwanted are these warnings:

```
==> Checking for packaging issues...
==> WARNING: Package contains reference to $srcdir
psp/lib/libharfbuzz.a
psp/lib/libharfbuzz-subset.a
==> WARNING: Package contains reference to $pkgdir
psp/lib/pkgconfig/harfbuzz-subset.pc
psp/lib/pkgconfig/harfbuzz.pc
```

Possibly something is not right during `prepare()` step?

**Comment by tpimh** [comment]
I have actually compiled my example and run it in an emulator. Here's the patch that I needed:
```diff
--- /usr/local/pspdev/psp/lib/cmake/harfbuzz/harfbuzzConfig.cmake
+++ harfbuzzConfig.cmake
@@ -59,8 +59,8 @@
 add_library(harfbuzz::harfbuzz STATIC IMPORTED)
 
 set_target_properties(harfbuzz::harfbuzz PROPERTIES
-  INTERFACE_INCLUDE_DIRECTORIES "/harfbuzz"
-  INTERFACE_LINK_LIBRARIES "Threads::Threads;/usr/local/pspdev/psp/lib/libfreetype.a"
+  INTERFACE_INCLUDE_DIRECTORIES "/usr/local/pspdev/psp/include/harfbuzz"
+  INTERFACE_LINK_LIBRARIES "/usr/local/pspdev/psp/lib/libfreetype.a;/usr/local/pspdev/psp/lib/libpng.a;/usr/local/pspdev/psp/lib/libz.a;/usr/local/pspdev/psp/lib/libbz2.a"
 )
 
 if(CMAKE_VERSION VERSION_LESS 2.8.12)
``

**Comment by sharkwouter** [comment]
Not sure if you're still working on this, but you can add the patch to this repo, no problem. For an example you could look at SDL2.

**Comment by tpimh** [comment]
@sharkwouter I will add the patch, I think it is still missing threads. It's just the example that is not using them, but anything that relies on threads would fail. It shouldn't be hard to add.

I can probably finalize this later today, and then try to link SDL2_ttf against harfbuzz.

**Comment by tpimh** [comment]
- [x] fix `Package contains reference to $pkgdir`
- [ ] fix `Package contains reference to $srcdir`
- [ ] fix `INTERFACE_INCLUDE_DIRECTORIES` and `INTERFACE_LINK_LIBRARIES` in generated cmake file

**Comment by sharkwouter** [comment]
I think the cmake file containing direct references to full paths is normal, since we have that for more packages but it does not seem to cause problems.

I've added a patch which fixes the issue with it trying to link to threads. Hope that helps. I would not try to fix references to the srcdir in .a files.

**Comment by sharkwouter** [comment]
/Can you add hardbuzz as a dependency for freetype? That is still missing.

**Comment by tpimh** [comment]
Yes, if this fix is okay with you, then it's fine with me as well. I think harfbuzz should also be added as a dependency for SDL2_ttf, I'll see if it builds and works.

**Review by sharkwouter — CHANGES_REQUESTED** [review]
Could you still add harfbuzz as dependency for sdl2_ttf?

**Comment by tpimh** [comment]
Sure, I can do it as a part of this pull request as well.

**Comment by tpimh** [comment]
Let's see if this test still fails when SDL2-ttf is built with harfbuzz: https://github.com/pspdev/psp-packages/issues/45#issuecomment-1286053609

File: harfbuzz/fix-cmake.patch [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
++++ new/CMakeLists.txt	2024-03-18 22:57:40.191313381 +0100
+@@ -116,7 +116,9 @@
+   find_package(Threads)
+   if (CMAKE_USE_PTHREADS_INIT)
+     add_definitions("-DHAVE_PTHREAD")
+-    list(APPEND THIRD_PARTY_LIBS Threads::Threads)
++    if (NOT PSP)
++      list(APPEND THIRD_PARTY_LIBS Threads::Threads)
++    endif()
+     list(APPEND PC_LIBS_PRIV -pthread)
```
Comment (sharkwouter) [inline harfbuzz/fix-cmake.patch:12]: I just did a test and this part also need to not be for PSP.
```suggestion
+    if (NOT PSP)
+      list(APPEND THIRD_PARTY_LIBS Threads::Threads)
+      ist(APPEND PC_LIBS_PRIV -pthread)
+    endif()
-     list(APPEND PC_LIBS_PRIV -pthread)
```
I did this suggestion by hand, so give me a second to test it

Comment (sharkwouter) [reply harfbuzz/fix-cmake.patch]: I updated the patch.

**Comment by sharkwouter** [comment]
I added a fix for psp-pkg-conf outputting something that does not work. I'm testing now.

**Comment by sharkwouter** [comment]
> @sharkwouter I would probably still add this patch to package step: https://github.com/pspdev/psp-packages/pull/115#issuecomment-1968801057. I know it's ugly, but I have no idea how to implement it better (and neither does the main harfbuzz dev: https://github.com/harfbuzz/harfbuzz/issues/4613#issuecomment-2006389772).

@tpimh the fix-cmake patch file already addresses this issue.

**Review by sharkwouter — CHANGES_REQUESTED** [review]
We have a dependency loop. Harfbuzz relies on Freetype2 and Freetype2 relies on Harfbuzz. Our build script cannot handle that.

What can we do with this? Does it have to be like this?

**Comment by tpimh** [comment]
Even after removing the circular dependency, the build system was not able to build the package without extra complications.

1. Duplicate dependencies are not supported. `sdl2-ttf` depends on `sdl2`, `freetype2` and `harfbuzz`, so the dependency list generated looks like `libpng libpng freetype2 libpng harfbuzz freetype2 sdl2 sdl2-ttf`. The first `libpng` builds fine, the second fails with "A package has already been built".
2. Preventing the duplicates to be added to the dependency list, only complicates things, because now the dependencies are not in the correct order: `libpng harfbuzz freetype2 sdl2 sdl2-ttf`. `harfbuzz` depends on `freetype2`, but is being built before it. Building it fails with "Could not resolve all dependencies".
3. Removing `freetype2` dependency from `sdl2-ttf` fixes the build. The dependency list is now `libpng freetype2 harfbuzz sdl2 sdl2-ttf`, the order is correct, because `freetype2` is added as dependency of `harfbuzz`.

**Comment by sharkwouter** [comment]
Thanks!

**Pull request merged**